docs/local_deployment.md

```diff
@@ -0,0 +1,121 @@
+# Local Deployment
+
+When using nested templates against Azure, they need to be publicly accessible. This is not easy to accomplish if you are developing on a laptop within a corporate network for example. To assist with this a local script has been setup that will perform the deployment for you.
+
+When the script stages are executed the following occurs:
+
+ - Does the resource group already exist
+   - If this is the first time the script has been run this will be false
+   - If this is not the first time, then a local file `.deploy` will be read and the iteration found. The resource group name is thus `<RESCOURCE_GROUP>-<ITERATION>`. Delete the resource group.
+ - Increment the iteration and write out to the `.deploy` file
+ - Determine the name of the resource group to create, again `<RESCOURCE_GROUP>-<ITERATION>`
+ - Upload files to the specified storage account and container
+ - Perform deployment into the newly created resource group
+
+## Limitations
+
+The are currently some few limitations on the tooling. These will be addressed
+
+ - The storage account and container must already exist
+ - The container must have public access set on it
+ - The deployment blocks until it is complete
+
+## Pre-requisites
+
+For the moment a storage account and container with public access needs to be configured.
+
+An Azure Service Principal Name stored in a credentials file. By default it will be looked for in `~/.azure/credentials`.
+
+## Configuration
+
+In order for deployments to work, a deployment configuration file is required. By default it will be found in the project root directory and is called `deploy.json`.
+This file should not be checked into source control and as such is part of the `.gitignore`.
+
+This file should contain the following information:
+
+| Attribute | Description | Example |
+|---|---|---|
+| resource_group.name | The name of the resource group to create to perform the deployment into | |
+| resource_group.location | Azure location for the resource group | |
+| resource_group.parameters_file | Path to the parameters file for the deployment | `.local/parameters.json` |
+| storage_account.name | Name of the storage account that the files are to be uploaded to | |
+| storage_account.container | Container within the storage account for the files | |
+| storage_account.group_name | Resource group that contains the storage account for uploads | |
+| dirs.working | The directory containing the built template files | `build/working` |
+
+### Upload
+
+The upload sub-command has the following syntax and options.
+
+_Syntax_
+
+`deploy.js upload <SUBSCRIPTION> [-s <STORAGE_ACCOUNT_NAME>] [-n <CONTAINER_NAME>] [-a <AZURE_CREDS_FILE>] [-G <STORAGE_ACCOUNT_RESOURCE_GROUP>]`
```

Comment:
Not a bad idea, but we are then playing with needing both Node and Ruby
I know we are in that situation today, but I went with Node for Azure Devops because that is what the build process uses and is cross platform
I may remove the Ruby side of things, and replicate in node so that it is all just node
The `package.json` file allows some automation which we are already using so I could put something in here that wraps up all the upload / deploy etc etc.